### PR TITLE
Added Getting Started notebook

### PR DESCRIPTION
**Pull request by Priyansi**
cc @trsvchn @vfdev-5 @sdesrozis

### CODE REVIEW

File: getting-started.ipynb [hunk 1]
```diff
@@ -0,0 +1,528 @@
...
+    "pip install pytorch-ignite\n",
+    "```\n",
+    "\n",
+    "1.  From `conda`\n",
+    "\n",
+    "``` shell\n",
+    "conda install ignite -c pytorch\n",
+    "```\n",
+    "\n",
+    "See [here](/docs/how-to-guides/installation) for other installation\n",
```
Comment (vfdev-5) [inline getting-started.ipynb:50]: Nit: maybe, absolute link here instead of relative as examples can be also used

File: getting-started.ipynb [hunk 2]
```diff
@@ -0,0 +1,528 @@
...
+    "Import the following:"
+   ],
+   "metadata": {}
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "source": [
+    "import torch\n",
+    "import torch.nn.functional as F\n",
```
Comment (vfdev-5) [inline getting-started.ipynb:69]: I was also wanted to get rid of the calls of functional part and use `nn.*` instead. Let's do that here.

File: getting-started.ipynb [hunk 3]
```diff
@@ -0,0 +1,528 @@
...
+    "\n",
+    "    def forward(self, x):\n",
+    "        x = F.relu(F.max_pool2d(self.conv1(x), 2))\n",
+    "        x = F.relu(F.max_pool2d(self.conv2_drop(self.conv2(x)), 2))\n",
+    "        x = x.view(-1, 320)\n",
+    "        x = F.relu(self.fc1(x))\n",
+    "        x = F.dropout(x, training=self.training)\n",
+    "        x = self.fc2(x)\n",
+    "\n",
+    "        return F.log_softmax(x, dim=-1)\n",
```
Comment (vfdev-5) [inline getting-started.ipynb:114]: Maybe, we can take a model from torchvision, slightly adapt few layers and do not bother with the complete definition etc

File: getting-started.ipynb (hunk 2 above)
Comment (Priyansi) [reply getting-started.ipynb]: Removed this.

File: getting-started.ipynb (hunk 1 above)
Comment (Priyansi) [reply getting-started.ipynb]: Changed to https://pytorch-ignite.netlify.app/docs/how-to-guides/installation

File: getting-started.ipynb (hunk 3 above)
Comment (Priyansi) [reply getting-started.ipynb]: Used a modified ResNet 18 along with changes in optimizer and loss function.

**Comment by Priyansi** [comment]
@vfdev-5 I have made the suggested changes. Could you please take a look again?

**Review by KickItLikeShika — COMMENTED** [review]
Thanks for the updates @Priyansi, I left a one comment.

File: getting-started.ipynb [hunk 4]
```diff
@@ -0,0 +1,630 @@
...
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "Saizk3heEcFz"
+      },
+      "source": [
+        "import torch\n",
+        "from torch import nn\n",
+        "from torch.optim import SGD\n",
```
[the remark this answers is not in the document]
Comment (Priyansi) [reply getting-started.ipynb]: Thanks for pointing that out. I have removed it now.

File: getting-started.ipynb [hunk 5]
```diff
@@ -0,0 +1,601 @@
...
+        "class Net(nn.Module):\n",
+        "\n",
+        "    def __init__(self, in_channels=1):\n",
+        "        super(Net, self).__init__()\n",
+        "        \n",
+        "        # Changed the output layer to output 10 classes instead of 1000 classes\n",
+        "        self.model = resnet18(num_classes=10)\n",
+        "\n",
+        "        # Changed the input layer to take grayscale images for MNIST instaed of RGB images\n",
+        "        self.model.conv1 = nn.Conv2d(1, 64, kernel_size=(7, 7), stride=(2, 2), padding=(3, 3), bias=False)\n",
```
Comment (vfdev-5) [inline getting-started.ipynb:162]: Please reduce `kernel_size=(7, 7)` to something smaller, .e.g 3x3

**Comment by vfdev-5** [comment]
@Priyansi thanks for the updates. I also thought about including code to setup logging to TensorBoard as pytorch has it as a submodule with a dependency ?

**Comment by Priyansi** [comment]
Sure I think it'd be a nice addition. I'll add the code.

File: getting-started.ipynb [hunk 6]
```diff
@@ -0,0 +1,601 @@
...
+      ],
+      "metadata": {
+        "id": "DDIW2zedEcF3"
+      }
+    },
+    {
+      "cell_type": "code",
+      "execution_count": null,
+      "source": [
+        "def get_data_loaders(train_batch_size, val_batch_size):\n",
```
Comment (sdesrozis) [inline getting-started.ipynb:192]: Why introduce this function rather than the plain code ?

Comment (Priyansi) [reply getting-started.ipynb:192]: This function was used in several MNIST ignite/examples. Although I can replace it with plain code!

Comment (Priyansi) [reply getting-started.ipynb]: Done!

File: getting-started.ipynb (hunk 5 above)
Comment (Priyansi) [reply getting-started.ipynb]: Done!

File: getting-started.ipynb [hunk 7]
```diff
@@ -0,0 +1,591 @@
...
+      "cell_type": "code",
+      "execution_count": 3,
+      "source": [
+        "import torch\n",
+        "from torch import nn\n",
+        "from torch.utils.data import DataLoader\n",
+        "from torchvision.datasets import MNIST\n",
+        "from torchvision.models import resnet18\n",
+        "from torchvision.transforms import Compose, Normalize, ToTensor\n",
+        "from tqdm import tqdm\n",
```
Comment (sdesrozis) [inline getting-started.ipynb:110]: Never used. 
However, the `ProgressBar` handler could be a nice feature to add, couldn't it ?

File: getting-started.ipynb [hunk 8]
```diff
@@ -0,0 +1,591 @@
...
+    {
+      "cell_type": "code",
+      "execution_count": 13,
+      "source": [
+        "@trainer.on(Events.EPOCH_COMPLETED)\n",
+        "def log_training_results(trainer):\n",
+        "    evaluator.run(train_loader)\n",
+        "    metrics = evaluator.state.metrics\n",
+        "    print(f\"Training Results - Epoch[{trainer.state.epoch}] Avg accuracy: {metrics['accuracy']:.2f} Avg loss: {metrics['nll']:.2f}\")\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:382]:
```suggestion
        "\n",
        "\n",
```

File: getting-started.ipynb [hunk 9]
```diff
@@ -0,0 +1,591 @@
...
+        "from torchvision.datasets import MNIST\n",
+        "from torchvision.models import resnet18\n",
+        "from torchvision.transforms import Compose, Normalize, ToTensor\n",
+        "from tqdm import tqdm\n",
+        "\n",
+        "from ignite.engine import Engine, Events, create_supervised_trainer, create_supervised_evaluator\n",
+        "from ignite.metrics import Accuracy, Loss\n",
+        "\n",
+        "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:524]:
```suggestion
        "\n",
        "\n",
```

File: getting-started.ipynb [hunk 10]
```diff
@@ -0,0 +1,591 @@
...
+        "}\n",
+        "evaluator = create_supervised_evaluator(model, metrics=val_metrics, device=device)\n",
+        "\n",
+        "# how many batches to wait before logging training status\n",
+        "log_interval=20\n",
+        "\n",
+        "@trainer.on(Events.ITERATION_COMPLETED(every=log_interval))\n",
+        "def log_training_loss(engine):\n",
+        "    print(f\"Epoch[{engine.state.epoch}] Loss: {engine.state.output:.2f}\")\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:569]:
```suggestion
        "\n",
        "\n",
```

File: getting-started.ipynb [hunk 11]
```diff
@@ -0,0 +1,591 @@
...
+        "@trainer.on(Events.EPOCH_COMPLETED)\n",
+        "def log_training_results(trainer):\n",
+        "    evaluator.run(train_loader)\n",
+        "    metrics = evaluator.state.metrics\n",
+        "    print(f\"Training Results - Epoch[{trainer.state.epoch}] Avg accuracy: {metrics['accuracy']:.2f} Avg loss: {metrics['nll']:.2f}\")\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:575]:
```suggestion
        "\n",
        "\n",
```

File: getting-started.ipynb [hunk 12]
```diff
@@ -0,0 +1,591 @@
...
+        "@trainer.on(Events.EPOCH_COMPLETED)\n",
+        "def log_validation_results(trainer):\n",
+        "    evaluator.run(val_loader)\n",
+        "    metrics = evaluator.state.metrics\n",
+        "    print(f\"Validation Results - Epoch[{trainer.state.epoch}] Avg accuracy: {metrics['accuracy']:.2f} Avg loss: {metrics['nll']:.2f}\")\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:581]:
```suggestion
        "\n",
        "\n",
```

File: getting-started.ipynb [hunk 13]
```diff
@@ -0,0 +1,591 @@
...
+        "    def __init__(self, in_channels=1):\n",
+        "        super(Net, self).__init__()\n",
+        "        \n",
+        "        # Changed the output layer to output 10 classes instead of 1000 classes\n",
+        "        self.model = resnet18(num_classes=10)\n",
+        "\n",
+        "        # Changed the input layer to take grayscale images for MNIST instaed of RGB images\n",
+        "        self.model.conv1 = nn.Conv2d(1, 64, kernel_size=(3, 3), stride=(2, 2), padding=(3, 3), bias=False)\n",
+        "\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:164]:
```suggestion
```

File: getting-started.ipynb [hunk 14]
```diff
@@ -0,0 +1,591 @@
...
+        "    def forward(self, x):\n",
+        "        return self.model(x)\n",
+        "\n",
```
Comment (trsvchn) [inline getting-started.ipynb:167]:
```suggestion
        "\n",
        "\n",
```

File: getting-started.ipynb [hunk 15]
```diff
@@ -0,0 +1,591 @@
...
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 10,
+      "source": [
+        "trainer = create_supervised_trainer(model, optimizer, criterion, device)\n",
+        "\n",
+        "val_metrics = {\n",
+        "    \"accuracy\": Accuracy(),\n",
+        "    \"nll\": Loss(criterion)\n",
```
Comment (trsvchn) [inline getting-started.ipynb:254]: Why do we use `nll` here? resnet output layer is linear, that's why we use crossentropy (logsoftmax+nll), can we call it nll in this case?
@vfdev-5

File: getting-started.ipynb [hunk 16]
```diff
@@ -0,0 +1,591 @@
...
+        "id": "4r_PUH1yEcF1"
+      }
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 5,
+      "source": [
+        "class Net(nn.Module):\n",
+        "\n",
+        "    def __init__(self, in_channels=1):\n",
```
Comment (trsvchn) [inline getting-started.ipynb:155]:
```suggestion
        "    def __init__(self):\n",
```
We do not use arg `in_channels`

**Review by trsvchn — COMMENTED** [review]
Thanks @Priyansi! I left some comments, mostly related to PEP8 conventions.

**Comment by Priyansi** [comment]
> @Priyansi thanks for the updates. I also thought about including code to setup logging to TensorBoard as pytorch has it as a submodule with a dependency ?

I suppose if we add all handlers available [here](https://pytorch.org/ignite/v0.4.5/generated/ignite.contrib.handlers.tensorboard_logger.html#ignite.contrib.handlers.tensorboard_logger.TensorboardLogger) the code will become bulky. I propose we only log optimizer's parameters (since we are logging losses and metrics via custom functions) and then link a how-to guide exploring other handlers available under tensorboard_logger. What are your thoughts on this?

**Comment by trsvchn** [comment]
> > @Priyansi thanks for the updates. I also thought about including code to setup logging to TensorBoard as pytorch has it as a submodule with a dependency ?
> 
> @vfdev-5 I suppose if we add all handlers available [here](https://pytorch.org/ignite/v0.4.5/generated/ignite.contrib.handlers.tensorboard_logger.html#ignite.contrib.handlers.tensorboard_logger.TensorboardLogger) the code will become bulky. I propose we only log optimizer's parameters (since we are logging losses and metrics via custom functions) and then link a how-to guide exploring other handlers available under tensorboard_logger. What are your thoughts on this?

What about storing train/eval results in a plain dicts: `train_history`, `eval_history` and plot the resulta at the end with matplotlib?

**Comment by vfdev-5** [comment]
@Priyansi let's do the same as here: https://labs.quansight.org/blog/2020/09/pytorch-ignite/
```python
from ignite.contrib.handlers import TensorboardLogger

# Define a Tensorboard logger
tb_logger = TensorboardLogger(log_dir="quick-start-mnist-output")

# Attach handler to plot trainer's loss every 100 iterations
tb_logger.attach_output_handler(
    trainer,
    event_name=Events.ITERATION_COMPLETED(every=100),
    tag="training",
    output_transform=lambda loss: {"batchloss": loss},
)

# Attach handler to dump evaluator's metrics every epoch completed
for tag, evaluator in [("training", train_evaluator), ("validation", evaluator)]:
    tb_logger.attach_output_handler(
        evaluator,
        event_name=Events.EPOCH_COMPLETED,
        tag=tag,
        metric_names="all",
        global_step_transform=global_step_from_engine(trainer),
    )
```

File: getting-started.ipynb (hunk 7 above)
Comment (Priyansi) [reply getting-started.ipynb]: Thanks for pointing that out. Added ProgressBar. However, I don't understand what you mean by 
> EDIT : Only 5 epochs, maybe let the code in a simple version should be more relevant.

File: getting-started.ipynb (hunk 15 above)
Comment (vfdev-5) [reply getting-started.ipynb:2647]: Right, thanks @trsvchn let's call it "loss"

**Comment by Priyansi** [comment]
> > > @Priyansi thanks for the updates. I also thought about including code to setup logging to TensorBoard as pytorch has it as a submodule with a dependency ?
> > 
> > 
> > @vfdev-5 I suppose if we add all handlers available [here](https://pytorch.org/ignite/v0.4.5/generated/ignite.contrib.handlers.tensorboard_logger.html#ignite.contrib.handlers.tensorboard_logger.TensorboardLogger) the code will become bulky. I propose we only log optimizer's parameters (since we are logging losses and metrics via custom functions) and then link a how-to guide exploring other handlers available under tensorboard_logger. What are your thoughts on this?
> 
> What about storing train/eval results in a plain dicts: `train_history`, `eval_history` and plot the resulta at the end with matplotlib?

@trsvchn I don't understand what you mean. Should I not use Tensorboard? Also what Ignite concepts can I introduce via this?

File: getting-started.ipynb [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+      "metadata": {
+        "id": "Sw90sOK9EcF9"
+      }
+    },
+    {
+      "cell_type": "code",
+      "execution_count": 11,
+      "source": [
+        "# How many batches to wait before logging training status\n",
+        "log_interval=100"
```
[the remark this answers is not in the document]
Comment (Priyansi) [reply getting-started.ipynb]: Thank you for the suggested changes. I have addressed them in my latest commit.

**Comment by trsvchn** [comment]
> @trsvchn I don't understand what you mean. Should I not use Tensorboard? Also what Ignite concepts can I introduce via this?

No, I'm ok with TB. I didn't get the context initially. I was thinking that the main intention was to show the training progress,

File: getting-started.ipynb (hunk 7 above)
Comment (sdesrozis) [reply getting-started.ipynb]: If you want to keep the code very simple, maybe ProgressBar is too much, especially regarding to the low number of epochs.

Comment (Priyansi) [reply getting-started.ipynb]: Got it. We can introduce them later in tutorials as it isn't a major Ignite concept and I would like to keep this notebook as short as possible.

**Comment by vfdev-5** [comment]
Git does not show the content of the notebook as json in the "Files"

1) Fix "Both trainer and evaluator objects"...
```python
train_evaluator = create_supervised_evaluator(model, metrics=val_metrics, device=device)
val_evaluator = create_supervised_evaluator(model, metrics=val_metrics, device=device)
```

> Both trainer and evaluator objects are instances of Engine - the main component of Ignite, which is essentially an abstraction over the training or validation loop.

We introduced `train_evaluator` and `val_evaluator` but below speaking about only `evaluator`.

2) Fix "We have defined separate evaluator engines for training and validation because they will serve different functions as we will see later in this tutorial"

This statement is not clear as later we do not see any particular reason why we have to create 2 evaluation engines. To make it more obvious let's add simple `Checkpoint` for the best model validated with `val_evaluator`.


3) Improve logging
```diff
@trainer.on(Events.ITERATION_COMPLETED(every=log_interval))
def log_training_loss(engine):
-     print(f"Epoch[{engine.state.epoch}] Loss: {engine.state.output:.2f}")
+     print(f"Epoch[{engine.state.epoch}], Iter[{engine.state.iteration}] Loss: {engine.state.output:.2f}")
```

**Comment by Priyansi** [comment]
> This statement is not clear as later we do not see any particular reason why we have to create 2 evaluation engines. T

I have mentioned here the reason however, I agree it's not that prominent:
We will use [`TensorboardLogger()`](https://pytorch.org/ignite/v0.4.5/generated/ignite.contrib.handlers.tensorboard_logger.html#ignite.contrib.handlers.tensorboard_logger.TensorboardLogger) to log trainer's loss, and training and validation metrics separately. This is why we defined two evaluator engines previously.

I will add `Checkpoint` to make it more obvious.

**Comment by vfdev-5** [comment]
@Priyansi thanks ! Just for your understanding why using TensorBoardLogger is not the reason to use 2 evaluators. Technically we can do this code:
```python
# Attach handler to dump evaluator's metrics after every epoch is completed
for tag, evaluator in [("training", train_evaluator), ("validation", val_evaluator)]:
    tb_logger.attach_output_handler(
        evaluator,
        event_name=Events.EPOCH_COMPLETED,
        tag=tag,
        metric_names="all",
        global_step_transform=global_step_from_engine(trainer),
    )
```
with a single evaluator like here: https://github.com/pytorch/ignite/discussions/2042

**Comment by Priyansi** [comment]
> @Priyansi thanks ! Just for your understanding why using TensorBoardLogger is not the reason to use 2 evaluators. Technically we can do this code:
> 
> ```python
> # Attach handler to dump evaluator's metrics after every epoch is completed
> for tag, evaluator in [("training", train_evaluator), ("validation", val_evaluator)]:
>     tb_logger.attach_output_handler(
>         evaluator,
>         event_name=Events.EPOCH_COMPLETED,
>         tag=tag,
>         metric_names="all",
>         global_step_transform=global_step_from_engine(trainer),
>     )
> ```
> 
> with a single evaluator like here: [pytorch/ignite#2042](https://github.com/pytorch/ignite/discussions/2042)

Thank you for sharing this! Yes, I did think about using just one `evaluator` engine however then we'll have to go through extra effort to create these tags so that training and validation metrics are logged separately. 
However, using `ModelCheckpoint` will definitely emphasize using two evaluators.

**Comment by vfdev-5** [comment]
@Priyansi 

This comment https://github.com/pytorch-ignite/examples/pull/2#issuecomment-884109082 point (1) is not fixed

Same for 
> If you need more control over your training and validation loops, you can create custom trainer and evaluator objects by wrapping the step logic in Engine :

and

> After an epoch ends during training, we can compute the training and validation metrics by running evaluator on train_loader and val_loader. 

This one: https://github.com/pytorch-ignite/examples/pull/2#issuecomment-884116701 also we need to fix.

Please, fix this today and we can merge this PR.

**Comment by Priyansi** [comment]
Sorry for the delay, all the concerns in [this comment](https://github.com/pytorch-ignite/examples/pull/2#issuecomment-884109082) have been addressed.

**Comment by vfdev-5** [comment]
> We will use TensorboardLogger() to log trainer's loss, and training and validation metrics separately. This is also why we defined two evaluator engines previously.

Let's remove: `This is also why we defined two evaluator engines previously.`

**Comment by vfdev-5** [comment]
Links like that `https://pytorch.org/ignite/v0.4.5/generated/ignite.contrib.handlers.tensorboard_logger.html#ignite.contrib.handlers.tensorboard_logger.TensorboardLogger` are too tightly coupled to the v0.4.5 version, let's use something like:
https://pytorch.org/ignite/generated/ignite.contrib.handlers.tensorboard_logger.html#ignite.contrib.handlers.tensorboard_logger.TensorboardLogger

**Comment by vfdev-5** [comment]
Let's use `https://pytorch-ignite.ai` instead of `https://pytorch-ignite.netlify.app`. I'll setup cname a bit later.

**Comment by Priyansi** [comment]
I fixed the links and reviewed the notebook again. Everything is okay from my side.

**Comment by vfdev-5** [comment]
@Priyansi code generator still points to netlify but we always should provide https://code-generator.pytorch-ignite.ai/

**Comment by vfdev-5** [comment]
Where we will provide colab badge to execute the content on Colab: here or on the website ?

**Comment by Priyansi** [comment]
> Where we will provide colab badge to execute the content on Colab: here or on the website ?

Website itself.

**Comment by vfdev-5** [comment]
@Priyansi please resolve the conflict and it is ready to land

**Comment by vfdev-5** [comment]
@Priyansi please update the params of the conv1 like:
```diff
-        self.model.conv1 = nn.Conv2d(
-            1, 64, kernel_size=(3, 3), stride=(2, 2), padding=(3, 3), bias=False
-        )
+       self.model.conv1 = nn.Conv2d(
+            1, 64, kernel_size=3, padding=1, bias=False
+        )
```

File: getting-started.ipynb [hunk 18]
```diff
@@ -0,0 +1,742 @@
...
+      ],
+      "metadata": {
+        "id": "O6uwwXO8EcF-"
+      }
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "After an epoch ends during training, we can compute the training and\n",
+        "validation metrics by running `evaluator` on `train_loader` and\n",
```
Comment (vfdev-5) [inline getting-started.ipynb:372]: @Priyansi please fix finally this `evaluator` with appropriate names.

**Comment by Priyansi** [comment]
Resolved in [commit](https://github.com/pytorch-ignite/examples/commit/4e4033dfb635af6366e6bc0ef7c6b3c69f982d48)

**Pull request merged**